### ISSUE TITLE
Improve code documentation

### ISSUE THREAD

**Issue by vassudanagunta** [gohugoio/hugo#3976]
I was unable to find documentation of Hugo's data model, code architecture and interfaces. 

The lack of such hinders contributions from people other than those named @bep. My observation is that all non @bep contributions are peripheral. 

Without such documentation, trying to implement anything other than peripheral changes feels like working with spaghetti code --no offense intended.  This is either due to my ignorance, or a state of code that is natural for beta versions. But having some kind of minimal documentation will not only increase contributions, it will force the data model and interface to be more coherent... Having to document something tends to have that effect.

**Comment by vassudanagunta** [comment]
To what degree is @spf13's nearly two-year old vision (Issue #3205) still relevant, still the future? Has it been rendered obsolete by a newer vision?

**Comment by bep** [comment]
@vassudanagunta I'll keep this issue open to track improvements in the documentation area, but please take questions and discussions at https://discourse.gohugo.io/ -- post a link to that discussion here, but continue the discussion over there.

For people coming here without that link I will shortly say this:

* For questions about Steve's visions, you will have to ask him, and I suggest you do it off-GitHub (I suspect he has "notification overload")
* I agree that parts of the Hugo source code can be challenging to follow. The `hugolib` package is too big, a result of convenience and laziness I guess. And the page/file handler chain has some not very useful abstractions that make it hard to to follow for no good reason (that chain is removed/simplified in my "bundler branch")
* But the "data model" is simple: A build contains one or more `Site` with one or more `Pages`.
* There is one `Build` method (1) that calls `process`, `assemble` and `render` for each `Site`.
* The above is everything in a nutshell and should not be too hard to follow if you really want to.
* We (currently) have no plugin support, so there are very few "paint by numbers" type of tasks.
* You have shown interest in the Black Friday upgrade. The rendering part was back when it was simply put into the `helpers` package. We need to do better now if we want to take advantage of the new features. But it is still a couple of functions (`RenderBytes`?) which should be possible to follow. Why don't you start with that and create a proposal for an elegantly designed package that does this right?
* This requires some considerable amount of work, of course.

But take all of that in other issues and on the discussion forum.
